### ISSUE TITLE
Import Plugin for Godot ?

### ISSUE THREAD

**Issue by Jeremi360** [Inochi2D/inochi2d#16]
Will/Can you made import plugin for Godot ?
Godot if you didn't know is Open Source 2D and 3D game engine: https://godotengine.org
I'm working on Rakugo a tools kit for crating Visual Novels, Point & Click Adventures and RPGs for Godot: https://rakugoteam.github.io
And I think your tool would work in with Godot/Rakugo.
There are D-lang bindings for Godot: https://github.com/godot-d/godot-d

**Comment by LunaTheFoxgirl** [comment]
I want to eventually get around to making versions for Godot and Unity, but it will first be when the spec is more complete. Of course other developers are free to implement the spec in their engine of choice in the meantime.

The main problem with supporting Inochi2D in Godot is the fact that we use stencil buffers, which Godot does not currently support; so a shader based workaround would have to be implemented in godot.

**Comment by LunaTheFoxgirl** [comment]
I also think that @you-win was working on a Godot port? But I'm not sure how far they got.

**Comment by Jeremi360** [comment]
> I want to eventually get around to making versions for Godot and Unity, but it will first be when the spec is more complete. Of course other developers are free to implement the spec in their engine of choice in the meantime.

Is great to hear that :D , I just joined your Patreon. 

> I also think that @you-win was working on a Godot port? But I'm not sure how far they got.

Okay I will check that.

**Comment by you-win** [comment]
I was/still am planning on a Godot port. My current focuses are (in order):

1. Finishing up the next minor release for [puppeteer](https://github.com/virtual-puppet-project/puppeteer)
2. Adding mediapipe support
3. Working on inochi2d/live2d in tandem (with some work already in progress [here](https://github.com/virtual-puppet-project/godot-cubism))

**Comment by LunaTheFoxgirl** [comment]
At this point Inochi2D is usable for vtubing use, so you should be able to start implementing the spec.

**Comment by LunaTheFoxgirl** [comment]
I've found a potential solution for adding Inochi2D support to godot through a plugin that adds a new node type. If it pans out the plugin will be made for Godot 4.

**Comment by Jeremi360** [comment]
> I've found a potential solution for adding Inochi2D support to godot through a plugin that adds a new node type. 
If it pans out the plugin will be made for Godot 4.

Is grate to hear that 😄

**Comment by GreatestFool** [comment]
I'm certainly interested in seeing a plugin of some sort.

**Comment by LunaTheFoxgirl** [comment]
> I'm certainly interested in seeing a plugin of some sort.

A plugin is planned to be implemented via GDExtensions, it will only be made for export targets that supports RenderingDevice, though.

**Comment by GreatestFool** [comment]
That's fine, I'm not sure how else you would implement this considering Godot proposal [3373](https://github.com/godotengine/godot-proposals/issues/3373) didn't really pan out. Maybe someday they'll expose everything and make it easier to implement without making a custom implementation ¯\\\_(ツ)_/¯

How much manual work does the planned implementation leave to users of the plugin? I'm assuming it'll be necessary to set up a custom shader or something of some sort? I haven't used the RenderingDevice functionality, so I don't really know much more than a little reading can teach.

**Comment by LunaTheFoxgirl** [comment]
> That's fine, I'm not sure how else you would implement this considering Godot proposal [3373](https://github.com/godotengine/godot-proposals/issues/3373) didn't really pan out. Maybe someday they'll expose everything and make it easier to implement without making a custom implementation ¯\_(ツ)_/¯

Yeah...

> How much manual work does the planned implementation leave to users of the plugin? I'm assuming it'll be necessary to set up a custom shader or something of some sort? I haven't used the RenderingDevice functionality, so I don't really know much more than a little reading can teach.

Ideally very little manual work, plan to do some trickery with adding Inochi2D models as an asset type to Godot via GDExtension so that some transformation can be done to the file on-load to convert a model in to a singular node.

**Comment by GreatestFool** [comment]
Sounds good. I'll just cross my fingers in the meantime, considering I likely won't be of any help lol. Good luck to you all, and thanks for making the project :)

**Comment by almondlynx** [comment]
Is it still in works? This would allow me to do so, so much

**Comment by LunaTheFoxgirl** [comment]
> Is it still in works? This would allow me to do so, so much

It's planned for 0.9, work is underways but I'll first go full in to development on the update after the holiday season (starting January) some of the components are already in the works, such as [numem](https://github.com/Inochi2D/numem) which is neccesary for integration with game engines, as D's GC was causing too many collissions with other memory managment strategies.

**Comment by BunkWire2X8** [comment]
Don't want to sound impatient, because I'm sure that this whole development process takes awhile, but I was wondering how far along this plugin's development has been going along?

Also, a bit related, will there also be built-in support for rendering Inochi rigs in 3D spaces? Been looking for the _longest_ time to find a way to render 2D skeletal systems in 3D without needing to resort to hacky viewport workarounds.

**Comment by LunaTheFoxgirl** [comment]
The progress is currently that the 0.9 rewrite has begun, but it's still quite far from ready to be integrated into a plugin.

**Comment by Vadimledyaev** [comment]
Hi, Is there any progress in porting the engine to Godot?

**Comment by LunaTheFoxgirl** [comment]
> Hi, Is there any progress in porting the engine to Godot?

There has been a tiny bit of progress on Inochi2D 0.9, but it's still quite far off for now.

0.8 is far too tied to OpenGL 3.2 to be used in games.

**Comment by LunaTheFoxgirl** [comment]
Update: I've managed to get godot-dlang working so that's one big hurdle overcome in getting Inochi2D running in Godot.

It still depends on me making more progress on the 0.9 SDK, but it's a starting point and it uses the gdextension system to auto-generate the bindings. So it should be relatively future proofed.

![Image](https://github.com/user-attachments/assets/796b05cf-4b03-44c7-acd2-15f44676d89c)

**Comment by mark-use-blender** [comment]
hi, i am working on a inochi2D parser in godot naively, is there any existing project that i can pull from or contribute to. 

BR,
Mark

**Comment by ngenthub** [comment]
I hope that inochi2d can be used with godot 3.x. That would be perfect for kind of game like visual novel.
but, i'm still waiting for Inochi2d to be usable with godot 4.x.
Thank you for this amazing tool.

**Comment by LunaTheFoxgirl** [comment]
Yeah; currently working on a big update for Inochi2D 0.8 to be the general sendoff for that version series; after that my focus will be shifted primarily to 0.9 and engine integration support.

The support is planned to be distributed as a GDExtension with precompiled builds for all the platforms godot supports officially. Said extension will be written in D like Inochi2D itself; which will ensure minimal overhead.

**Comment by LunaTheFoxgirl** [comment]
Update; the work on this has now started in earnest, due to difficulties with OpenGL in cross platform settings I have today ripped out all the hardcoded OpenGL code. I'll progressively be rebuilding Inochi2D to be a bring your own renderer library and I'll be testing using Godot and Unity. Once I have the basics working in godot I'll release the source for the integration GDExtension.

Do note that initial versions will likely be poorly optimised as I plan to do this transition pragmatically by starting out with a stupid draw list system; over time improving its smarts.

**Comment by DearFox** [comment]
> Update; the work on this has now started in earnest, due to difficulties with OpenGL in cross platform settings I have today ripped out all the hardcoded OpenGL code. I'll progressively be rebuilding Inochi2D to be a bring your own renderer library and I'll be testing using Godot and Unity. Once I have the basics working in godot I'll release the source for the integration GDExtension.
> 
> Do note that initial versions will likely be poorly optimised as I plan to do this transition pragmatically by starting out with a stupid draw list system; over time improving its smarts.

I'm very happy to hear this, I can't wait to start testing how everything works.

The initial implementation will only support rendering via Vulkan for godot ? Or when you talk about OpenGL, you mean how Inochi2D works?

**Comment by LunaTheFoxgirl** [comment]
> > Update; the work on this has now started in earnest, due to difficulties with OpenGL in cross platform settings I have today ripped out all the hardcoded OpenGL code. I'll progressively be rebuilding Inochi2D to be a bring your own renderer library and I'll be testing using Godot and Unity. Once I have the basics working in godot I'll release the source for the integration GDExtension.
> > Do note that initial versions will likely be poorly optimised as I plan to do this transition pragmatically by starting out with a stupid draw list system; over time improving its smarts.
> 
> I'm very happy to hear this, I can't wait to start testing how everything works.
> 
> The initial implementation will only support rendering via Vulkan for godot ? Or when you talk about OpenGL, you mean how Inochi2D works?

Inochi2D itself is moving away from doing any rendering itself essentially. All it'll do is give you "draw lists" that you yourself have to send to the GPU and graphics API of your choice. I plan to make a godot plugin which does this, initially just with the new rendering infrastructure in godot, but later I wish to possibly also support a subset of Inochi2D in the legacy mode.

**Comment by mark-use-blender** [comment]
> hi, i am working on a inochi2D parser in godot naively, is there any existing project that i can pull from or contribute to.
> 
> BR, Mark

its almost been a year since all things started (for me) I've bee mostly throwing everything at the wall and see if anything sticks, the import (translation) script is already on my github but it abuses backframe copy to render masks and alpha properly, so, single digit FPS, but if you don't care about masks and alpha, it can to animations by shoving vertex coordinates in to the polygon2Ds, resulting in a 60 FPS animation (which is really curse but it works)

**Comment by LunaTheFoxgirl** [comment]
Godot plugin is now in development, the code is available in `modules/godot/`

Build instructions are here, though right now it only loads a model, still need to write the rendering code
https://github.com/Inochi2D/inochi2d?tab=readme-ov-file#inochi2d-godot

**Comment by LunaTheFoxgirl** [comment]
I'm having some issues interacting with RenderingDevice from GDExtension currently, still need to track down why, but once that's figured out I should be able to get Inochi2D working relatively quickly at this point.
